include/vara/Feature/ConstraintParser.h

```diff
@@ -0,0 +1,492 @@
+      }
+      return {ConstraintToken(ConstraintToken::ConstraintTokenKind::NOT), 1};
+    case '=':
+      if (Str.startswith("=>")) {
+        return {ConstraintToken(ConstraintToken::ConstraintTokenKind::IMPLIES),
+                2};
+      }
+      return {ConstraintToken(ConstraintToken::ConstraintTokenKind::EQUAL), 1};
+    case '>':
+      if (Str.startswith(">=")) {
+        return {ConstraintToken(
+                    ConstraintToken::ConstraintTokenKind::GREATER_EQUAL),
+                2};
+      }
+      return {ConstraintToken(ConstraintToken::ConstraintTokenKind::GREATER),
+              1};
+    case '<':
+      if (Str.startswith("<->") || Str.startswith("<=>")) {
+        return {
+            ConstraintToken(ConstraintToken::ConstraintTokenKind::EQUIVALENT),
+            3};
+      } else if (Str.startswith("<=")) {
+        return {
+            ConstraintToken(ConstraintToken::ConstraintTokenKind::LESS_EQUAL),
+            2};
+      }
+      return {ConstraintToken(ConstraintToken::ConstraintTokenKind::LESS), 1};
+    default:
+      return {ConstraintToken(ConstraintToken::ConstraintTokenKind::ERROR,
+                              Str.take_front().str()),
+              1};
+    }
+  }
+
+  static ResultTy munchNumber(const llvm::StringRef &Str) {
+    auto Munch = Str.take_while([](auto C) { return llvm::isDigit(C); });
+    return {ConstraintToken(ConstraintToken::ConstraintTokenKind::NUMBER,
+                            Munch.str()),
+            Munch.size()};
+  }
+
+  static ResultTy munchIdentifier(const llvm::StringRef &Str) {
+    auto Munch =
+        Str.take_while([](auto C) { return llvm::isAlnum(C) || C == '_'; });
+    return {ConstraintToken(ConstraintToken::ConstraintTokenKind::IDENTIFIER,
+                            Munch.str()),
+            Munch.size()};
+  }
+
+  std::string Cnt;
+};
+
+//===----------------------------------------------------------------------===//
+//                               ConstraintParser
+//===----------------------------------------------------------------------===//
+
+class ConstraintParser {
+public:
+  explicit ConstraintParser(std::string Cnt)
+      : TokenList(ConstraintLexer(std::move(Cnt)).tokenize()) {}
+
+  std::unique_ptr<Constraint> buildConstraint() { return parseConstraint(); }
+
+private:
+  [[nodiscard]] const ConstraintToken &peek() const {
+    return TokenList.front();
+  }
+
+  [[nodiscard]] ConstraintToken next() {
+    auto Token = TokenList.front();
+    TokenList.pop_front();
+    return Token;
+  }
+
+  bool consume(const ConstraintToken::ConstraintTokenKind Kind) {
+    if (TokenList.front().getKind() == Kind) {
+      TokenList.pop_front();
+      return true;
+    }
+    return false;
+  }
+
+  std::unique_ptr<Constraint>
+  parseConstraint(int NestingLevel = 0,
+                  ConstraintToken::PrecedenceTy Precedence =
+                      ConstraintToken::MaxPrecedence) {
+    if (auto LHS = parseUnaryConstraint(NestingLevel)) {
```

Comment:
You don't really need the if here. If LHS is a nullptr while(LHS) would skip it anyhow.